### ISSUE TITLE
Water Segmentation Models

### ISSUE THREAD

**Issue by bsahil29** [cc-ai/kdb#62]
We are planning to work on segmentation models that have been designed specifically for the water scenario : 
- Train the network
- Assert author performances
- Comparison with deeplab
- (optional) Fine-tuning on our flooded House dataset

The approach used is described in [Single Image Water Detection](http://openaccess.thecvf.com/content_ECCV_2018/papers/Xiaofeng_Han_Single_Image_Water_ECCV_2018_paper.pdf). The net is designed to detect water on asphalt and country road using reflection unit.

The paper code : [Github Link](https://github.com/Cow911/SingleImageWaterHazardDetectionWithRAU)

**Comment by gcosne** [comment]
**Experiment 1 : Training on puddles & Infering on flooded scenes**
- GPU : Tesla k80 12GB 
- Batch size : 1
- Training epochs : 50K  
- Training dataset : [puddle dataset ](https://cloudstor.aarnet.edu.au/plus/s/oSeR8zogqzaXN6X)
- Training time : 12 hours

***Remark :*** We trained the network with the same parameters as given by the author of the paper except for the epoch number (50K vs 60K). Infering on flooded images while training on puddle dataset doesn't work well. 

See qualitative results : **puddle_detection** &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;  &nbsp; &nbsp; &nbsp;&nbsp; &nbsp; &nbsp;                         **failed_case_flood_segmenter**

<img src="https://user-images.githubusercontent.com/25148268/58177745-ef097080-7c72-11e9-8011-64f354feddf7.JPG" height="185"><img src="https://user-images.githubusercontent.com/25148268/58176394-29bdd980-7c70-11e9-8531-e23f9a4aace8.JPG" height="185">

**Conclusion :** I suggest we wait for the author to answer and deliver the supplementary materials where there is a detailed analysis on training time, robustness to over-fitting etc..

If we agree to spend more time testing this method, we should train it over our own flooded scene.

**Comment by gcosne** [comment]
Here is a [Link ](https://github.com/gcosne/SingleImageWaterHazardDetectionWithRAU/blob/master/README.md) to the fork.

**Comment by gcosne** [comment]
**Experiment 2 : Initializing with the puddle model and training over 50 flooded segmented images**

- Training iteration : 250
- Training dataset : [Flood_images](https://drive.google.com/open?id=1zkFqHf-CoPenqmR8zmQboUFfUKm9Y0CW)
- Training time : 10 min


***Remark 1 :*** This experiment aim to evaluate the ability of such a network to segment flood, results are promising for a fine-tuning.

We could try to extract soft segmentation mask and combine it with other methods with fuzzy theory.

***Remark 2:***  The model hasn't been shown images without water so it is biased, one may want to train the latter with such images.

| Qualitative evaluation| on testing images| |
|:-------------------------:|:-------------------------:|:-------------------------:|
|<img width="1604" alt="screen shot 2017-08-07 at 12 18 15 pm" src="https://user-images.githubusercontent.com/25148268/58337365-a6d58400-7e13-11e9-80db-77b7515ec7a4.png"> |  <img width="1604" alt="screen shot 2017-08-07 at 12 18 15 pm" src="https://user-images.githubusercontent.com/25148268/58337366-a6d58400-7e13-11e9-83cf-657075c05463.png">|<img width="1604" alt="screen shot 2017-08-07 at 12 18 15 pm" src="https://user-images.githubusercontent.com/25148268/58337367-a6d58400-7e13-11e9-8a73-51f5c5ddbd79.png">|
|<img width="1604" alt="screen shot 2017-08-07 at 12 18 15 pm" src="https://user-images.githubusercontent.com/25148268/58337368-a6d58400-7e13-11e9-99eb-fd0c25688cc7.png">  |  <img width="1604" alt="screen shot 2017-08-07 at 12 18 15 pm" src="https://user-images.githubusercontent.com/25148268/58337369-a6d58400-7e13-11e9-89e1-8f502b7e1ab4.png">|<img width="1604" alt="screen shot 2017-08-07 at 12 18 15 pm" src="https://user-images.githubusercontent.com/25148268/58337370-a6d58400-7e13-11e9-9b80-16cc75bd2506.png">|
|<img width="1604" alt="screen shot 2017-08-07 at 12 18 15 pm" src="https://user-images.githubusercontent.com/25148268/58337372-a6d58400-7e13-11e9-9177-d50ede1497f3.png">  |  <img width="1604" alt="screen shot 2017-08-07 at 12 18 15 pm" src="https://user-images.githubusercontent.com/25148268/58337373-a6d58400-7e13-11e9-8460-d7f58de75035.png">|<img width="1604" alt="screen shot 2017-08-07 at 12 18 15 pm" src="https://user-images.githubusercontent.com/25148268/58337374-a76e1a80-7e13-11e9-9a2e-6739ced82b8d.png">|


| Qualitative evaluation| on training images| |
|:-------------------------:|:-------------------------:|:-------------------------:|
|<img width="1604" alt="screen shot 2017-08-07 at 12 18 15 pm" src="https://user-images.githubusercontent.com/25148268/58339308-4e07ea80-7e17-11e9-929e-0a7ca5c0d774.png"> |  <img width="1604" alt="screen shot 2017-08-07 at 12 18 15 pm" src="https://user-images.githubusercontent.com/25148268/58339309-4e07ea80-7e17-11e9-8d78-6ab870db9d47.png">|<img width="1604" alt="screen shot 2017-08-07 at 12 18 15 pm" src="https://user-images.githubusercontent.com/25148268/58339310-4e07ea80-7e17-11e9-8bca-f44d9a9eefb8.png">|
|<img width="1604" alt="screen shot 2017-08-07 at 12 18 15 pm" src="https://user-images.githubusercontent.com/25148268/58339311-4e07ea80-7e17-11e9-92a1-eeb936c993ee.png">  |  <img width="1604" alt="screen shot 2017-08-07 at 12 18 15 pm" src="https://user-images.githubusercontent.com/25148268/58339314-4ea08100-7e17-11e9-9c9f-fd8e2afd7c1d.png">|<img width="1604" alt="screen shot 2017-08-07 at 12 18 15 pm" src="https://user-images.githubusercontent.com/25148268/58339316-4ea08100-7e17-11e9-8bc3-00c2156b70e5.png">|
|<img width="1604" alt="screen shot 2017-08-07 at 12 18 15 pm" src="https://user-images.githubusercontent.com/25148268/58339318-4ea08100-7e17-11e9-94e9-08c08405ef84.png">  |  <img width="1604" alt="screen shot 2017-08-07 at 12 18 15 pm" src="https://user-images.githubusercontent.com/25148268/58339319-4ea08100-7e17-11e9-86d8-35d4926f8d51.png">|<img width="1604" alt="screen shot 2017-08-07 at 12 18 15 pm" src="https://user-images.githubusercontent.com/25148268/58339320-4ea08100-7e17-11e9-9f0b-c1d586cdaa55.png">|

**Conclusion :** 
As we can see the network is still in early stage of training (it does not fit very well the training data).

**Comment by gcosne** [comment]
**To do** : 
- Quantitative evaluation train/test 
- Fine tune really (not training the entire model)

**Comment by gcosne** [comment]
**Update :**  Reading this [github issue](https://github.com/NVlabs/SPADE/issues/17) 

Starting from the observation that [GauGAN ](https://blogs.nvidia.com/blog/2019/03/18/gaugan-photorealistic-landscapes-nvidia-research/) can draw pretty realistic water, I look deeper into their paper and github issue to find which was the semantic segmentation model used in their experiment with flickr landscape dataset. Following this [github issue](https://github.com/NVlabs/SPADE/issues/17), they used [DeepLab v2 on COCO-stuffs-164k ](https://github.com/kazuto1011/deeplab-pytorch).

**Evaluating Results**
Evaluating on our dataset of manually annotated 210 images of flood, without the _CRF  post processing_, results appear very good, the mean IOU is about 79% with some example that totally fail.

A good way of using it would be to process our entire dataset and remove manually every image with less than 50% of water covered. If we do that post evaluation on the 210 images, we keep 88 % of them and increase the IOU to 86%. 

**Good Results**:
![output_crf_deeplab](https://user-images.githubusercontent.com/25148268/60675756-214de680-9e4b-11e9-8b12-e3b87e151486.png)

![im_output](https://user-images.githubusercontent.com/25148268/60675479-7c330e00-9e4a-11e9-8e4b-d1d48e9b8e00.png)

**Worst Results**:
![image (3)](https://user-images.githubusercontent.com/25148268/60675858-66721880-9e4b-11e9-8cc7-9b90e555ae83.png)

![image (2)](https://user-images.githubusercontent.com/25148268/60675811-40e50f00-9e4b-11e9-80d1-53cebe950748.png)

The algorithm fails to recognize water, when it recognize the road, it also happens when the image is in heavy rain condition (few centimeters over the road and we still see the lane)

**Comment by gcosne** [comment]
Here is the implementation of the selected algorithm: https://github.com/cc-ai/deeplab-pytorch/

<blockquote><img src="https://avatars3.githubusercontent.com/u/48456149?s=400&v=4" width="48" align="right"><div><img src="https://github.githubassets.com/favicon.ico" height="14"> GitHub</div><div><strong><a href="https://github.com/cc-ai/deeplab-pytorch">cc-ai/deeplab-pytorch</a></strong></div><div>PyTorch implementation of DeepLab v2 on COCO-Stuff  - cc-ai/deeplab-pytorch</div></blockquote>
